### ISSUE TITLE
Array#includes check with enum

### ISSUE THREAD

**Issue by deecewan** [facebook/flow#4246]
Hey,

So, I'm pretty sure that this will have a fairly logical solution, but I've been trawling through bugs and stack overflow and, I can't find a) a solution, and more importantly b) an explanantion.

I'm trying to validate that an item is part of an `enum`.

```js
type StringEnum = 'first_option' | 'second option' | 'third option'

const options: Array<StringEnum> = ['first_option', 'second option', 'third option']

function checkInEnum(item: string): boolean {
     if (options.includes(item)) {
        console.log('it\'s a valid option')
        return true;
     }
     return false;
}

checkInEnum('some item')
checkInEnum('first_option')
```

However, in the `includes` call, flow gets mad because `item` is not a valid subset of `StringEnum`. Which I (probably) know, and I'm trying to check for.

Is there a nice way to fix this?

[Flow Try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDKGATgJYB2A5gKIUCuAtmALxgDkUZJAzhgPpwcGMnArswAHw488AYzEATMEJFiJ09hgAW3ZatHj0CinxXDDPAFxgAgiRIBDLAB5i5anSYA+VmADanNx8ghbqADQy8krmauKRWrok+mHiALroUPQUcnFgctryANYAkhRejAAUZBh4jDZ8HlQAlDYARnBwMHiOFGAA3qhgw8NkUGCVBmI8AHSUcjD0ing81bWMzc0DQyO7JjxdeDPwVJXsNQA67DxgjmAAbo4wZClx7M07u8MkeBj0JH1SPQ8ABuT7DAC+4LAPz+ALAUCesjBUNQBWKZQqZwOjAINTq7zRhTkpXKDCqQV4Aim4maQA)

**Comment by niieani** [comment]
Use: `checkInEnum(item: StringEnum)` instead. This way you'll catch incorrect usage of `checkInEnum` build-time, rather than in runtime. This kinda makes the function not necessary. 

Why would you want to do the check otherwise? Are you receiving the value from a backend?

**Comment by deecewan** [comment]
that was actually a terrible example, sorry.

a better function signature would be:

```js
function normalizeValue(item: string) {
  if (options.includes(item)) {
    return item;
  }
  return item.split(' ').join(',');
}
```

I need to do different things based on whether the value is in the array or not, which is why I need to check.  The value is 'kind of' user provided. If it's in the list, it's been set using a toggle, otherwise, it's been set using a text field.

**Comment by niieani** [comment]
Ok, that makes sense. Not sure how to type this without casting `item` to `any` within the `.includes` call. In TypeScript you'd simply cast `item as StringEnum` in the call to `includes`, but in Flow that's not possible.

**Comment by madbence** [comment]
this sounds like a bug to me (?), `.include` is defined on `$ReadOnlyArray<T>`, so it should accept anything that is a supertype of `T`.

**Comment by popham** [comment]
It feels like a tautological "explanation," but this doesn't work because `string` is not a subtype of your enum and the signature of `includes` looks like so:

```js
includes(searchElement: T, fromIndex?: number): boolean;
```

By the spec this is too restrictive: http://www.ecma-international.org/ecma-262/7.0/#sec-array.prototype.includes. That said, I prefer it this way because I'm certain that I can conjure up use cases that benefit from it as-is (and I prefer false negatives to false positives).

Casting is the quick work around, but I would probably remodel the data.

**Comment by ryanscottaudio** [comment]
@popham how would you fix something like this?

```type StringEnum = 'first_option' | 'second option' | 'third option'
const options = ['first_option', 'second option', 'third option'];

function normalizeValue(item: string): StringEnum {
  return options.includes(item)
  	?  item
  	: 'first_option';
}```

Seems to me that this should work.

**Comment by niieani** [comment]
@ryanscottaudio it cannot work, because we cannot statically infer the contents of `options`. Even though it's a `const`, the contents of the Array might have changed in runtime. 

Here's kind of a workaround if you don't want casting: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDKGATgJYB2A5gKIUCuAtmALxgDkUZJAzhgPpwcGMnArswAHw488AYzEATQcNHipHDAAtuyoSLHsA3KgUU+YfWp6swAbU7c+Kg+IA0M+Updr2H9tq6PoYAuiaoivIwAIYkBFD0FHKuYJRyMPSRPAAUZBh4jABcYHzk1ACUxQBGcHAweNEUYACkclryANY22ahgqfnMLEMcXLwCVoZS0nkFrMPssmZ6qpOS0wNzbAE6JMuu7Kjl6AlJKRRwJIzRMGQAXngAatf0eLkDxaWUVBrEZbQMjEqRFIXzoTDAAG9emA4hh6CQmmkMlk3gUjn0+gB+foFaF9YoTcx2AAMYVQAF8gA

**Comment by ryanscottaudio** [comment]
@niieani what if it was a `$ReadOnlyArray`?

**Comment by niieani** [comment]
@ryanscottaudio yeah, that should work. All of the `$ReadOnlyArray` methods could be typed much more narrowly, including `includes(x)`.

**Comment by ryanscottaudio** [comment]
@niieani it doesn't work with $ReadOnlyArray: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDKGATgJYB2A5gKIUCuAtmALxgDkUZJAzhgPpwcGMnArswAHw488AYzEATQcNHipHDAAtuyoSLHsA3KgUU+YfWp4AuMABIASngCGigPIUYWAIIkSLlgAPMTk1HRMAHysYADanNx8KgbiADQy8krJauzp7Nq62YYAuiaoivIwLiQEUPQUcilglHIw9BU8ABRkGHiMdnxhVACUdgBGcHAwrhRgAKRyWvIA1jxgnahglqpiPAB0LW0d3b2Mw6jnqHUNTRRwJIwuMGQAXngAak-0eCd9A6SUKgaUKAiJnOwg8IMZgAb02YBqGHoJFmh3aeC6PT65y2WwA-M1TvCtnYrLtYgAGUqoAC+QA. I also don't get how your workaround works; what if someone did `options.unshift('not_part_of_string_enum')`?

**Comment by niieani** [comment]
@ryanscottaudio yes, I know it doesn't work with $ReadOnlyArray. That's why I wrote it "should work", not that it works 😆 . Workaround works because you explicitly declare that the declared `include` function can verify the input is only one of the 3 strings, nothing else.

**Comment by ryanscottaudio** [comment]
@niieani hahaha sorry, i thought you meant it should work if you put it in there. for the workaround though, `options[0]` wouldn't be a StringEnum if you unshifted something else into `options`, and that has nothing to do with `includes`; `normalizeValue` is supposed to return a StringEnum regardless of if `includes` returns `true` or `false`, and if `options[0]` isn't a StringEnum because something got unshifted into it beforehand, then `normalizeValue` won't return a StringEnum if `includes(item)` returns false

**Comment by Hypnosphi** [comment]
I ran into exactly the same issue as in https://github.com/facebook/flow/issues/4246#issuecomment-333607634
This suggestion to use `.find` turned out to be quite useful:
https://github.com/facebook/flow/issues/2982#issuecomment-265808159

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDKGATgJYB2A5gKIUCuAtmALxgDkUZJAzhgPpwcGMnArswAHw488AYzEATQcNHipHDAAtuyoSLHsA3KgUU+YfWp6swAbU7c+Kg+IA0M+Updr2H9tq6PoYAuiaoUPQUcq5gFHAkjACGMGQAXngAain0eAAUZBh4jABcYHzk1ACUZcSVtAzMAN6oYGAkeBj0JBSWqmI8AHRcFIp5VmKsAHx9sSzzYIXFVVLSE+Z2AAxhqAC+QA

**Comment by noppa** [comment]
The type definition of Array.includes has been changed at some point and OP's code no longer errors. This issue can be closed.

FWIW I think the original behaviour was better and the current type definition allows some easy to make mistakes to go unnoticed, as demonstrated [in a similar TypeScript issue](https://github.com/microsoft/TypeScript/issues/26255#issuecomment-479557916), but I guess that's a subject for another time and another issue.
